### PR TITLE
Avoid recursive calls in BatchFetchQueue

### PR DESCRIPTION
**Pull request by bahusoid**
My suggested refactoring to simplify code reading (see my comment [here](https://github.com/nhibernate/nhibernate-core/pull/1633#discussion_r259292997) for details)

### CODE REVIEW

File: src/NHibernate/Async/Engine/BatchFetchQueue.cs [hunk 1]
```diff
@@ -74,7 +74,7 @@ public Task<object[]> GetCollectionBatchAsync(ICollectionPersister collectionPer
 			foreach (KeyValuePair<CollectionEntry, IPersistentCollection> me in map)
 			{
 				cancellationToken.ThrowIfCancellationRequested();
-				if (await (ProcessKeyAsync(me)).ConfigureAwait(false))
+				if (await (ProcessKeyAndCheckCacheAsync(me)).ConfigureAwait(false))
```
Comment (bahusoid) [inline src/NHibernate/Async/Engine/BatchFetchQueue.cs]: Awaiting for each cache key check also doesn't look effective... Though not sure yet how to improve it

Comment (bahusoid) [reply src/NHibernate/Async/Engine/BatchFetchQueue.cs]: Inline is the answer :)

**Comment by maca88** [comment]
The same refactoring should be done also for `GetEntityBatch` method.

File: src/NHibernate/Engine/BatchFetchQueue.cs [hunk 2]
```diff
@@ -276,7 +276,7 @@ bool CheckCacheAndProcessResult()
 				{
 					for (var j = 0; j < collectionKeys.Count; j++)
 					{
-						if (ProcessKey(collectionKeys[indexes[j]].Key))
+						if (ProcessKey(collectionKeys[indexes[j]].Key) == true)
```
Comment (hazzik) [inline src/NHibernate/Engine/BatchFetchQueue.cs]: Here `batchableCache == null` so `ProcessKey` never returns `null`.

File: src/NHibernate/Engine/BatchFetchQueue.cs [hunk 3]
```diff
@@ -287,7 +287,7 @@ bool CheckCacheAndProcessResult()
 					var results = AreCached(collectionKeys, indexes, collectionPersister, batchableCache, checkCache);
 					for (var j = 0; j < results.Length; j++)
 					{
-						if (!results[j] && ProcessKey(collectionKeys[indexes[j]].Key, true))
+						if (!results[j] && ProcessKey(collectionKeys[indexes[j]].Key, true) == true)
```
Comment (hazzik) [inline src/NHibernate/Engine/BatchFetchQueue.cs]: Here `ignoreCache = true` so ProcessKey never returns `null`.

File: src/NHibernate/Engine/BatchFetchQueue.cs [hunk 4]
```diff
@@ -459,7 +459,7 @@ bool CheckCacheAndProcessResult()
 				{
 					for (var j = 0; j < entityKeys.Count; j++)
 					{
-						if (ProcessKey(entityKeys[indexes[j]].Key))
+						if (ProcessKey(entityKeys[indexes[j]].Key) == true)
```
Comment (hazzik) [inline src/NHibernate/Engine/BatchFetchQueue.cs]: Here `batchableCache == null` so `ProcessKey` never returns `null`.

File: src/NHibernate/Engine/BatchFetchQueue.cs [hunk 5]
```diff
@@ -470,7 +470,7 @@ bool CheckCacheAndProcessResult()
 					var results = AreCached(entityKeys, indexes, persister, batchableCache, checkCache);
 					for (var j = 0; j < results.Length; j++)
 					{
-						if (!results[j] && ProcessKey(entityKeys[indexes[j]].Key, true))
+						if (!results[j] && ProcessKey(entityKeys[indexes[j]].Key, true) == true)
```
Comment (hazzik) [inline src/NHibernate/Engine/BatchFetchQueue.cs]: Here `ignoreCache == true` so `ProcessKey` never returns `null`.

**Comment by hazzik** [comment]
Done also for `GetEntityBatch` method.

**Pull request merged**